pkg/kubelet/images/image_manager.go

```diff
@@ -164,14 +233,37 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
 		m.backOff.Next(backOffKey, m.backOff.Clock.Now())
 
 		msg, err := evalCRIPullErr(container, imagePullResult.err)
-		return "", msg, err
+		return imageRef, msg, err
 	}
 	m.podPullingTimeRecorder.RecordImageFinishedPulling(pod.UID)
 	imagePullDuration := time.Since(startTime).Truncate(time.Millisecond)
 	m.logIt(ref, v1.EventTypeNormal, events.PulledImage, logPrefix, fmt.Sprintf("Successfully pulled image %q in %v (%v including waiting). Image size: %v bytes.",
 		container.Image, imagePullResult.pullDuration.Truncate(time.Millisecond), imagePullDuration, imagePullResult.imageSize), klog.Info)
 	metrics.ImagePullDuration.WithLabelValues(metrics.GetImageSizeBucket(imagePullResult.imageSize)).Observe(imagePullDuration.Seconds())
 	m.backOff.GC()
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) && m.pullImageSecretRecheckPeriod.Duration > 0 {
+		func() {
+			m.lock.Lock()
+			defer m.lock.Unlock()
+			if imagePullResult.pullCredentialsHash == "" {
+				// successful pull no auth hash returned, auth was not required so we should reset the hashmap for this
+				// imageref since auth is no longer required for the local image cache, allowing use of the ImageRef
+				// by other pods if it remains cached and pull policy is PullIfNotPresent
+				delete(m.ensureSecretPulledImages, imageRef)
```

Comment:
I kinda feel like pullCredentialsHash == "" should be handled in a separate mechansim. As in: if we remove images with an empty credentials hash from this map, then we treat the first pull after kubelet restart and first pull after recheck period as `pullCredentialHash == "'"`. Instead, can this be skipped or even add a "" entry in to the maps, and then we treat not present in map as always recheck?

Reply:
If the image needs no hash at first and then later needs auth, this would be a problem.

If recheck is enabled, the recheck for no auth image seems to be needed as well.

Reply:
https://github.com/kubernetes/kubernetes/compare/7a120ec8850900a419db335db37cf7063cba3b3f..51f8822fe18a3b044c76d866a765c8109b32b403 the latest update .

Reply:
@pacoxu 
What do you think about storing an "anonymous" entry instead of starting over.. when anonymous is successful.. If we go that route we can check for known anonymous success first then if not in the list check for each hash.. With duration checking enabled can also recheck anonymous..

Reply:
iow isEnsuredBySecret() needs to be improved to isEnsured() // By Secret || Anonymous  .. when anonymous worked for a particular registry that image is now in cache and thus any private pull with auth is moot, at least until another policy says recheck it now..  

I'm taking the presumption that we are not going to check if an image was pulled from registry a or registry b.. the image in the k8s.io image cache is still the same bytes either way..